monitor.go

```diff
@@ -0,0 +1,36 @@
+package nuki
+
+import (
+	"context"
+	"encoding/hex"
+
+	"github.com/go-ble/ble"
+)
+
+type StateChangeHandler func(context.Context, *Client)
+
+// MonitorStateChanges listens to advertisements from the given devices and triggers the callback
+// once their state changes. The client - which must be paired with ClientIdTypeBridge - can then
+// call ReadStates to fetch the new state and reset the flag.
+func MonitorStateChanges(ctx context.Context, clb StateChangeHandler, clients ...*Client) error {
+	h := func(a ble.Advertisement) {
+		for _, c := range clients {
+			if a.Addr().String() == c.addr.String() {
+				_, stateChanged := ParseAdvertisement(a)
+				if stateChanged && !c.stateChanged {
+					go clb(ctx, c)
+				}
+				c.stateChanged = stateChanged
```

Comment:
Is there any specific reason why this needs to be in a struct? Even if you want to decouple this from the client, it could as well be a local variable in that method.

There's another motivation to keep it in the client though: Automatically re-enabling the notifications when `readStates()` has been called. The only other way I can see this happing is to assume that the callback will always call `readStates()`, and therefore to reset the flag when the callback has finished. But if the callback decides to postpone the states retrieval, the lock will keep sending the flag and the library will think it's a new event...

Reply:
No your are right: there is no need for a own struct. Which brings me to another point, though:

The current implementation will "observe" clients which are configured at the start of monitoring. What if your application will create another client after the monitoring is starting. Therefore you have to stop, and start the monitoring again with the new client. Maybe it is a better approach to make an own struct which can hold the "observed" clients, where you can add/remove clients at runtime:

```
func NewMonitor() *monitor {
	return &monitor{
		watchedClients:      map[string]*Client{},
		watchedClientsMutex: sync.RWMutex{},
	}
}

func (m *monitor) ObserveClient(clients ...*Client) {
	m.watchedClientsMutex.Lock()
	defer m.watchedClientsMutex.Unlock()

	for _, client := range clients {
		m.watchedClients[client.addr.String()] = client
	}
}

func (m *monitor) StopObserveClient(clients ...*Client) {
	m.watchedClientsMutex.Lock()
	defer m.watchedClientsMutex.Unlock()

	for _, client := range clients {
		delete(m.watchedClients, client.addr.String())
	}
}

// MonitorStateChanges listens to advertisements from the given devices and triggers the callback
// once their state changes. The client - which must be paired with ClientIdTypeBridge - can then
// call ReadStates to fetch the new state and reset the flag.
func (m *monitor) MonitorStateChanges(ctx context.Context, clb StateChangeHandler) error {
	h := func(a ble.Advertisement) {
		m.watchedClientsMutex.RLock()
		c := m.watchedClients[a.Addr().String()]
		m.watchedClientsMutex.RUnlock()

		if c != nil {
			_, stateChanged := ParseAdvertisement(a)
			if stateChanged && !c.stateChanged {
				go clb(ctx, c)
			}
			c.stateChanged = stateChanged
		}
	}
	return ble.Scan(ctx, true, h, nil)
}
```